### PR TITLE
sql: implement the PostgreSQL wire protocol

### PR DESCRIPTION
**Pull request by tamird**
Out for early review per @petermattis.

Transactions are not yet implemented, and authentication is incomplete.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3218)

### CODE REVIEW

**Comment by tamird** [comment]
OK, this should be passing now. Note that the last commit represents the last outstanding bit of work (implement type checking for subqueries, or equivalent).

Nevertheless, this is ready for review.

**Comment by tamird** [comment]
OK, subqueries should work now as well.

**Comment by petermattis** [comment]
Review status: 0 of 29 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[rpc/codec/conn.go, line 36 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vWGN5UOK-8sYnOSAx:-K3vWGN5UOK-8sYnOSAy:-1181639181) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/rpc/codec/conn.go#L36)):*
Let's revert this for now.

---

*[server/server.go, line 221 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vWvVpV6i0ENQYPuKR:-K3vWvVpV6i0ENQYPuKS:235780252) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/server/server.go#L221)):*
Perhaps add a `pgaddr` flag. 

---

*[sql/driver/wire.proto, line 99 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vXVJgGQZsKVgCwfLv:-K3vXVJgGQZsKVgCwfLw:1698636128) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/driver/wire.proto#L99)):*
Any reason for changing the field id to `3`?

---

*[sql/logic_test.go, line 466 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vYCWoYuYw_D9mwJTu:-K3vYCWoYuYw_D9mwJTv:1453120591) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/logic_test.go#L466)):*
What does `%q` do if the argument is not a string/[]byte?

---

*[sql/logic_test.go, line 480 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vYQZGM2LqXzxeeCzh:-K3vYQZGM2LqXzxeeCzi:-1497524547) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/logic_test.go#L480)):*
Can you add a comment explaining what this is for? You're converting the `[]byte` to a string, but that's different than `fmt.Sprint()`?

---

*[sql/parser/values.go, line 36 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vYhDf3SD53zIUnTbF:-K3vYhDf3SD53zIUnTbG:-143753303) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/parser/values.go#L36)):*
Any reason for the changes here?

---

*[sql/scan.go, line 530 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vZWDyE0h83-EUaaEC:-K3vZWDyE0h83-EUaaED:-1688613307) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/scan.go#L530)):*
Type checking twice? Probably deserves a comment to explain that we have to type check the normalized expression to determine its output type.

---

*[sql/testdata/datetime, line 20 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vZvGtKwBKVQmn_9Fg:-K3vZvGtKwBKVQmn_9Fh:790011992) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/testdata/datetime#L20)):*
Why did this change? Pgwire might be sending the dates differently, but they still arrive as `time.Time`, right?

---

*[sql/testdata/txn, line 77 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3v_862btq4MMTCq30g:-K3v_862btq4MMTCq30h:2145507833) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/testdata/txn#L77)):*
Why is this being output? Are we outputting all of the results over pgwire or only the last one?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218)*

**Comment by petermattis** [comment]
Can you include the output of the new `Select1` variations?

---

Review status: 0 of 29 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-:-K3v_gQuwEH44NsWGSQd:1401747124)*

**Comment by tamird** [comment]
Review status: 0 of 28 files reviewed at latest revision, 9 unresolved discussions.

---

*[sql/driver/wire.proto, line 99 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vXVJgGQZsKVgCwfLv:-K3vanEZAQmwfNbNrZOq:529426274) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/driver/wire.proto#L99)):*
Done goofed. Reverted.

---

*[sql/logic_test.go, line 466 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vYCWoYuYw_D9mwJTu:-K3vas-CE7qo6toffzan:53268199) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/logic_test.go#L466)):*
actually, I'm not sure.

---

*[sql/logic_test.go, line 480 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vYQZGM2LqXzxeeCzh:-K3vblk_iKZgP3p9F5Pj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/logic_test.go#L480)):*
Done.

---

*[sql/parser/values.go, line 36 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vYhDf3SD53zIUnTbF:-K3vbmj4TinxfIDwz-pi:700048134) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/parser/values.go#L36)):*
Nah, just seemed simpler. It's in a separate commit.

---

*[sql/scan.go, line 530 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vZWDyE0h83-EUaaEC:-K3vce5sBYtiQaoNR2lR:1256157153) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/scan.go#L530)):*
Turns out it's possible to type check just once. Changed.

---

*[sql/testdata/datetime, line 20 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vZvGtKwBKVQmn_9Fg:-K3vch5x2jccZlGtKc0H:428150391) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/testdata/datetime#L20)):*
They were previously arriving as `driver.Date`, not `time.Time`

---

*[sql/testdata/txn, line 77 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3v_862btq4MMTCq30g:-K3vcoqzSaU5dxYKwMeS:1806070135) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/testdata/txn#L77)):*
That's the behaviour change (see one of the commit messages). pgwire is outputting them all.

---

*[rpc/codec/conn.go, line 36 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vWGN5UOK-8sYnOSAx:-K3va_6zAhhgCi0B60xr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/rpc/codec/conn.go#L36)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218)*

**Comment by tamird** [comment]
Sure. For now, here they are:
```
benchmark                    old ns/op     new ns/op     delta
BenchmarkBank-4              1046735       787929        -24.73%
BenchmarkSelect1_HTTP-4      490991        484219        -1.38%
BenchmarkSelect1_HTTPS-4     470506        519195        +10.35%
BenchmarkSelect1_RPC-4       121440        130514        +7.47%
BenchmarkSelect1_RPCS-4      142458        142702        +0.17%
```

---

Review status: 0 of 28 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-:-K3vpZFC_p4G4mi3Cd2H:-1500686396)*

**Comment by petermattis** [comment]
LGTM

What about the `_{PGWire,Postgres,MySQL}` variants? I know there won't be a before number, but the relative values of the 3 are interesting to make sure we didn't lose any of the optimizations from my earlier work.

---

Review status: 0 of 28 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/logic_test.go, line 466 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vYCWoYuYw_D9mwJTu:-K3vvXEZ-0PP0XiQMpo_:1777226107) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/logic_test.go#L466)):*
I'd use `%v` for the non-string types. 

---

*[sql/parser/values.go, line 36 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vYhDf3SD53zIUnTbF:-K3vvg_vInVT6xwdC0rR:-233092556) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/parser/values.go#L36)):*
Using the buffer is probably faster. Fine for now, though.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-:-K3vvEJAjXjQR5lZn-nM:-1854349368)*

**Comment by tamird** [comment]
Here's `cat new.txt`
```
go test -tags ''  -i ./sql
go test -tags ''  -run - -bench . ./sql -timeout 5m 
PASS
BenchmarkBank-4            	    2000	    787929 ns/op
BenchmarkSelect1_HTTP-4    	    3000	    484219 ns/op
BenchmarkSelect1_HTTPS-4   	    2000	    519195 ns/op
BenchmarkSelect1_RPC-4     	   10000	    130514 ns/op
BenchmarkSelect1_RPCS-4    	   10000	    142702 ns/op
BenchmarkSelect1_PGWire-4  	   10000	    142844 ns/op
BenchmarkSelect1_Postgres-4	   20000	     66301 ns/op
BenchmarkSelect1_MySQL-4   	   20000	     84914 ns/op
ok  	github.com/cockroachdb/cockroach/sql	14.594s
```

---

Review status: 0 of 28 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/logic_test.go, line 466 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K3vYCWoYuYw_D9mwJTu:-K3wifjJvQUdjY_LQIwG:-701012646) ([raw file](https://github.com/cockroachdb/cockroach/blob/66ca88ba4b8dc001ed841ce1da8230854453ab60/sql/logic_test.go#L466)):*
Changed to `%#v`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-:-K3wcnfeRGB_X9ywmT2M:861368759)*

**Comment by petermattis** [comment]
Hmm, I thought `PGWire` in my branch was as fast at `Postgres`. Wonder what happened. Oh well, we can figure it out after this is merged.

---

Review status: 0 of 28 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-:-K3z9IlbibbtEMZUsPAO:485568321)*

**Comment by tamird** [comment]
This is currently blocked on https://github.com/lib/pq/pull/408 and https://github.com/lib/pq/pull/409

**Comment by bdarnell** [comment]
LGTM

**Comment by tamird** [comment]
Confirmed this passes the bigger logic_test suite. Merging.

**Comment by tschottdorf** [comment]
LGTM

---

Reviewed 3 of 3 files at r1, 11 of 11 files at r2, 14 of 24 files at r3, 5 of 6 files at r4, 1 of 1 files at r5, 3 of 5 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[GLOCKFILE, line 26 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K48vyf15L669QnyMix5:-K48vyf15L669QnyMix6:-1405431518) ([raw file](https://github.com/cockroachdb/cockroach/blob/354c207732770b456c8f01d9335f2d8e46ff783c/GLOCKFILE#L26)):*
mysql?

---

*[sql/logic_test.go, line 478 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K492uaqOt900z1iArYE:-K492uaqOt900z1iArYF:-1589132955) ([raw file](https://github.com/cockroachdb/cockroach/blob/354c207732770b456c8f01d9335f2d8e46ff783c/sql/logic_test.go#L478)):*
s/out/our/

---

*[sql/pgwire/server.go, line 114 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K48z-AvkLcY3Cs1TMHE:-K48z-AvkLcY3Cs1TMHF:-473445218) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5acbb56f1372a85ded425db7081d67f473860a4/sql/pgwire/server.go#L114)):*
could pass `readBuffer` in to save an alloc.

---

*[sql/pgwire/v3.go, line 122 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K48zi02qS-Kq2W5EFoG:-K48zi02qS-Kq2W5EFoH:-113732341) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5acbb56f1372a85ded425db7081d67f473860a4/sql/pgwire/v3.go#L122)):*
can't the usage of `readBuf` be optimized? It looks like it's always written until it's full and then it'll allocate a new slice. Shouldn't it be able to re-use the existing buffer at some point (I was looking at the code in `readBuf.reset()`)?

---

*[rpc/codec/conn.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K4921DoTFheiBAmjExi:-K4921DoTFheiBAmjExj:678148092) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5acbb56f1372a85ded425db7081d67f473860a4/rpc/codec/conn.go#L36)):*
why?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-:-K493k67V3B2MiMMmJFL:676834233)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[GLOCKFILE, line 26 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K48vyf15L669QnyMix5:-K494gdKZsKymj8ZU_pC:618035931) ([raw file](https://github.com/cockroachdb/cockroach/blob/354c207732770b456c8f01d9335f2d8e46ff783c/GLOCKFILE#L26)):*
Used in benchmarks.

---

*[sql/logic_test.go, line 478 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K492uaqOt900z1iArYE:-K494nOD_Z4dFWl0z2z8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/354c207732770b456c8f01d9335f2d8e46ff783c/sql/logic_test.go#L478)):*
Done.

---

*[sql/pgwire/server.go, line 114 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K48z-AvkLcY3Cs1TMHE:-K4964bOt2eQPdns3N5V:260243866) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5acbb56f1372a85ded425db7081d67f473860a4/sql/pgwire/server.go#L114)):*
Wouldn't that not save anything since `readBuffer` would be copied by value and still allocated along with the rest of the `v3Conn`?

---

*[sql/pgwire/v3.go, line 122 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K48zi02qS-Kq2W5EFoG:-K495Hh7g8cLnHwIMYHR:191487616) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5acbb56f1372a85ded425db7081d67f473860a4/sql/pgwire/v3.go#L122)):*
Not sure what you mean. Seems to me that `reset()` allocates quite rarely, only when the buffer isn't big enough to hold a message of size `size`. What am I missing?

---

*[rpc/codec/conn.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218#-K4921DoTFheiBAmjExi:-K494vpS3w0USiQFXmAH:1344556940) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5acbb56f1372a85ded425db7081d67f473860a4/rpc/codec/conn.go#L36)):*
reviewable is messing with you, this code didn't change in this PR.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3218)*

File: sql/pgwire/encoding.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	msg []byte
+	tmp [4]byte
+}
+
+// reset sets b.msg to exactly size, attempting to use spare capacity
+// at the end of the existing slice when possible and allocating a new
+// slice when necessary.
+func (b *readBuffer) reset(size int) {
+	if b.msg != nil {
+		b.msg = b.msg[len(b.msg):]
```
Comment (tschottdorf) [inline sql/pgwire/encoding.go]: @tamird re: reviewable. Doesn't this mean that we never re-use anything of the existing buffer? So we'd alloc 4096 initially, use it up across multiple calls to `reset(.)` and then allocate another (presumably) 4096.

Comment (tamird) [inline sql/pgwire/encoding.go]: ah, yeah, this never reuses anything. I agree, this can be optimized.

**Pull request merged**